### PR TITLE
refactor!: update the lib to TS and require named parameters

### LINKED ISSUE #27: rewrite code in TS
Hey! How about Idea to rewrite code in TS? I can do it (already done actually)

### PR DESCRIPTION
**Pull request by dnim**
Resolves #27 

I still a bit didn't get how is better to update doc generation. Will be very glad for any help or comments

### CODE REVIEW

**Comment by coston** [comment]
Hey @dnim! Thank you for opening a PR. Before I review:
- update the PR to pass CI
- minimize the changes to only what is necessary for the feature. There are a lot of extra files, a package manager change, and directory changes that are unnecessary.

**Comment by dnim** [comment]
@coston hey! 
can you please take a look on this error? https://app.netlify.com/sites/json-to-csv-export/deploys/62ff2c0ee8a7ea00075ace02
I never worked before with `mdx-go` and `netlify`, so maybe it's something obvious for you. Already spent couple of hours on this. 

If no - I'll take a look deeper. 
Thanks!

**Comment by coston** [comment]
@dnim There are many changes, like bumping mdx-go a major version, `yarn` to `npm`, and `/dist` to `/built` that are unnecessary for your PR. 

Please simplify the PR to the minimal essentials for:
1. Updating the lib to TS
2. changed signature 
3. Testing ❤️
4. Docs update

**Comment by dnim** [comment]
@coston should be better now!

File: migration.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
+# Migration from 1.x.x to 2.x
+
+## Reason for migration
```
Comment (coston) [inline migration.md]:
```suggestion
## Breaking Changes
```

File: migration.md [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+New version was written in TS and a bit changed the arguments list to make it more convenient and flexible
```
Comment (coston) [inline migration.md]:
```suggestion
Named parameters are now required for passing custom `data`, `filename`, and `delimeter` props. 
```

Comment (coston) [reply migration.md]: How filenames are handled should be mentioned too.

File: src/index.ts [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+  link.click()
+  document.body.removeChild(link)
+}
+
+const getFilename = (providedFilename?: string): string => {
+  return providedFilename
+    ? providedFilename.toLowerCase().endsWith('.csv')
+      ? providedFilename
+      : `${providedFilename}.csv`
+    : 'export.csv'
```
Comment (coston) [inline src/index.ts]: Please no nested ternaries.

Comment (coston) [inline src/index.ts]: Just set `export.csv` at the top (like the delimeter) for visibility.

File: src/index.ts [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+  data: any[]
+  filename?: string
+  delimiter?: string
+  headers?: string[]
+}
+
+const CSV_FILE_TYPE = 'text/csv;charset=utf-8;'
+const csvDownload = ({
+  data,
+  filename, // default to 'export.csv'
```
Comment (coston) [inline src/index.ts]:
```suggestion
  filename = 'export.csv',
```

Comment (coston) [reply src/index.ts]: See my comments on filename below.

File: migration.md (hunk 2 above)
Comment (coston) [reply migration.md]: Removal of IE msSaveBlob support should be mentioned.

File: src/index.ts [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+    )
+    return
+  }
+
+  const headerKeys = Object.keys(items[0])
+
+  const columnNames = headers ?? headerKeys
+  const csv = items.map((row) =>
+    headerKeys
+      .map((fieldName) => JSON.stringify(row[fieldName] ?? ''))
```
Comment (coston) [inline src/index.ts]: Zero handling needs to be restored https://github.com/coston/json-to-csv-export/pull/29/files

File: test/index.spec.ts [hunk 6]
```diff
@@ -0,0 +1,48 @@
+import csvDownload from '../src/index';
+import mockData from './__mocks__/mockData';
+
+describe("csvDownload", () => {
```
Comment (coston) [inline test/index.spec.ts:4]: Thank you for this!

File: src/index.ts [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+  link.download = fileName
+  link.style.display = 'none'
+  document.body.appendChild(link)
+  link.click()
+  document.body.removeChild(link)
+}
+
+const getFilename = (providedFilename?: string): string => {
+  return providedFilename
+    ? providedFilename.toLowerCase().endsWith('.csv')
```
Comment (coston) [inline src/index.ts]: I was on the fence about this feature. It's not really needed, but it's nice. Consider using a regex instead, as I think it's faster than the toLowerCase + endsWith combo. 
```suggestion
    ? /csv$/i.test(providedFilename)
```

File: readme.md [hunk 8]
```diff
@@ -36,20 +39,44 @@ Or load from a CDN:
...
+    name: "Linell Humpherston",
+    ip: "82.225.151.150"
+  }
+]
+
+const dataToConvert = {
+  data: ipAddressesData,
+  filename: 'ip_addresses_report',
+  delimiter: ',',
+  headers: ['IP', "Full Name", "# Address"]
```
Comment (coston) [inline readme.md]:
```suggestion
  headers: ['IP', "Full Name", "IP Address"]
```

File: package.json [hunk 9]
```diff
@@ -18,33 +18,22 @@
   "license": "MIT",
   "scripts": {
     "dev": "mdx-go docs && npm run watch",
-    "docs": "mdx-go build docs -d public",
-    "build": "babel -d dist src --copy-files",
-    "clean": "rm -rf dist && mkdir dist",
+    "docs": "npm run build && mdx-go build docs -d public",
+    "build": "tsc",
+    "clean": "rm -rf dist",
     "watch": "babel -w -d dist src --copy-files",
```
Comment (coston) [inline package.json]: Please update this one to watch src changes when running dev

**Review by coston — CHANGES_REQUESTED** [review]
Looks great! Some suggestions and minor changes requested.

**Review by coston — APPROVED** [review]
Very nice, @dnim! I will make a couple of [nit] tweaks after merging. LGTM! 🚀

**Pull request merged**